### PR TITLE
Convert cudf::merge to use device_uvector instead of device_vector

### PR DESCRIPTION
**Pull request by harrism**
Converts `cudf::merge` to use `device_uvector`.

Contributes to #7287

Performance is improved: 

```
Comparing /home/mharris/rapids/cudf/cpp/build/merge_before.json to /home/mharris/rapids/cudf/cpp/build/merge_after.json
Benchmark                                          Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------
Merge/pow2tables/2/manual_time                  -0.0831         -0.0791             0             0             0             0
Merge/pow2tables/4/manual_time                  -0.0924         -0.0930             1             1             1             1
Merge/pow2tables/8/manual_time                  -0.0813         -0.0808             4             3             4             3
Merge/pow2tables/16/manual_time                 -0.1172         -0.1170             9             8             9             8
Merge/pow2tables/32/manual_time                 -0.0790         -0.0790            19            18            20            18
Merge/pow2tables/64/manual_time                 -0.0757         -0.0757            46            43            46            43
Merge/pow2tables/128/manual_time                -0.0730         -0.0730           111           103           111           103
```

### CODE REVIEW

File: cpp/include/cudf/detail/merge.cuh [hunk 1]
```diff
@@ -80,11 +80,11 @@ struct tagged_element_relational_comparator {
   __device__ weak_ordering compare(index_type lhs_tagged_index, index_type rhs_tagged_index) const
     noexcept
   {
-    side l_side = thrust::get<0>(lhs_tagged_index);
-    side r_side = thrust::get<0>(rhs_tagged_index);
+    side const l_side = thrust::get<0>(lhs_tagged_index);
```
Comment (karthikeyann) [inline cpp/include/cudf/detail/merge.cuh:83]: TIL, `thrust::get<0>()` works on `thrust::pair` too.

File: cpp/benchmarks/string/factory_benchmark.cu [hunk 2]
```diff
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#include "rmm/cuda_stream_view.hpp"
```
Comment (karthikeyann) [inline cpp/benchmarks/string/factory_benchmark.cu:17]: `<>` preferred.
```suggestion
#include <rmm/cuda_stream_view.hpp>
```

**Review by karthikeyann — APPROVED** [review]
LGTM.

**Review by rgsl888prabhu — APPROVED** [review]
Looks good,  have to update copyright

**Comment by harrism** [comment]
@karthikeyann @rgsl888prabhu please don't approve when changes are required, to prevent it getting merged before the fixes are made.

File: cpp/src/merge/merge.cu [hunk 3]
```diff
@@ -143,6 +145,12 @@ void materialize_bitmask(column_view const& left_col,
   CHECK_CUDA(stream.value());
 }
 
+struct side_index_generator {
+  side _side;
+
+  __device__ index_type operator()(size_type i) const noexcept { return index_type{_side, i}; }
```
Comment (karthikeyann) [inline cpp/src/merge/merge.cu:151]: Does this lambda work? (with trailing-return-type)
`[] __device__(auto i)  -> index_type { return index_type{_side, i}; }`

**Review by karthikeyann — CHANGES_REQUESTED** [review]

File: cpp/src/merge/merge.cu (hunk 3 above)
Comment (harrism) [reply cpp/src/merge/merge.cu:151]: I'm not going to experiment with that since the functor is already made.

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**